### PR TITLE
fix: resolve hub station route_variants mismatch causing no trains to display (#380)

### LINKED ISSUE #49: front end allows unroutable station combinations (although blocked by backend)
The front end allows users to select sequences like:
Southgate → Leicester Square → Northern → Bank

but this is not possible - Leicester Square and Bank are not on the same branch of the Northern Line.

this is correctly detected by the backend. But the frontend should not let the user select Bank.

### LINKED ISSUE #380: no trains on any line from hub stations
Image

also happens at waterloo, liverpool street - i think any hub. If selecting "bayswater" as the starting station, paddington or liverpool street dont show up as "circle line" stations either.

the complete steps are:
  1. Create new database
  2. check out main
  3. run {{baseUrl}}/api/v1/admin/tfl/build-graph -> it works
  4. call {{baseUrl}}/api/v1/admin/tfl/build-graph again -> it doesnt work.

the front end code does not change between steps 3 and 4. This is clearly an issue relating to the rebuild.

what's different between first build and second build:

First build:
- Database is empty
- Stations are created fresh via _create_new_station()
- Connections are created

Second build:
- Stations already exist in database
- Stations are updated via _update_existing_station()
- Existing connections are soft-deleted
- New connections are created

## core differential diagnosis

Looking at the specific API responses, when it's working: `/api/v1/tfl/lines` returns an object with `route_variants` as `null`. When it stops working, `route_variants` is populated.

Other API responses are identical. 

## goals and approach

Unacceptable solutions include:
- do NOT try and send deduplicated/hub IDs to the front end. This lookup/dereferencing is supposed to happen in the backend already.
- do NOT Add child_tfl_ids field to StationResponse that maps hub codes to their underlying NaPTAN IDs

Goals:

- identify why route_variants are null after first build
- identify why front end is unable to correctly use the route_variants. Should it even care about them (see #49 - not implemented)

### PR DESCRIPTION
**Pull request by mnbf9rca**
## Summary

Fixes #380 - Hub stations (Paddington, Bank, Liverpool Street, etc.) now correctly show available trains after database rebuilds.

Closes #49 

## Root Cause Analysis

Two bugs were identified and fixed:

### Bug #1: Cache Key Mismatch
**Problem**: `build_station_graph()` invalidated cache key `"lines"` but `fetch_lines()` actually uses `"lines:modes:dlr,elizabeth-line,overground,tube"`. After first build, stale cached data with `route_variants=null` was served until cache expired.

**Fix**: Updated cache invalidation to use the correct key format with modes.

### Bug #2: Station ID Mismatch  
**Problem**: 
- `route_variants.routes[].stations` contained raw NaPTAN IDs (e.g., `"940GZZLUPAC"` for Paddington Underground)
- Frontend received deduplicated stations with hub codes as `tfl_id` (e.g., `"HUBPAD"`)
- Frontend's `getNextStations()` couldn't match hub stations → no trains displayed

**Fix**: Store both raw and canonical versions of route_variants.

## Solution Architecture

Implemented a "normalize on read" pattern:

1. **`route_variants`** - Preserved authoritative TfL data with raw NaPTAN IDs (used by backend validation)
2. **`route_variants_canonical`** - New field with hub codes translated for API responses (used by frontend)

This approach:
- ✅ Preserves raw TfL data for auditing/debugging
- ✅ No breaking changes to backend validation logic
- ✅ Frontend receives matching IDs for hub stations
- ✅ Follows existing codebase patterns

### Database Changes
- Added `route_variants_canonical` JSON column to `lines` table
- Migration: `55ac4b0569a1_add_route_variants_canonical_to_lines.py`

### New Helper Functions
**`backend/app/helpers/station_resolution.py`**:
- `build_naptan_to_canonical_map()` - Maps NaPTAN IDs to canonical station IDs (hub codes or self)
- `translate_route_variants_to_canonical()` - Translates station lists in route variants

### Service Updates
**`backend/app/services/tfl_service.py`**:
- Fixed cache key mismatch in `build_station_graph()`
- Added `_compute_canonical_route_variants()` to compute canonical versions after stations are loaded
- Imports new translation helpers

### API Response
**`backend/app/schemas/tfl.py`**:
- Updated `LineResponse` to read from `route_variants_canonical` field via `validation_alias`
- Maintains backward compatibility - field still named `route_variants` in API response

### New Tests (8 added)
**`backend/tests/test_station_resolution_helpers.py`**:
- `TestBuildNaptanToCanonicalMap` (4 tests)
  - Hub station mapping
  - Standalone station mapping
  - Mixed stations
  - Empty list handling
- `TestTranslateRouteVariantsToCanonical` (7 tests)
  - Hub station translation
  - Fallback behavior for missing stations
  - None/empty input handling
  - Route structure preservation

### Test Results
- ✅ **1,523 backend tests passed** (no regressions)
- ✅ **48 station resolution helper tests passed** (including 8 new tests)
- ✅ **All pre-commit hooks passed** (ruff, mypy, formatting)

## Verification Steps

1. ✅ Database migration applied successfully
2. ✅ All tests passing (1,523 tests, 0 failures)
3. ✅ Pre-commit hooks passing

### Manual Verification
To verify the fix works:
1. Clear database and rebuild graph: `POST /api/v1/admin/tfl/build-graph`
2. Rebuild again (second build): `POST /api/v1/admin/tfl/build-graph`
3. Check `/api/v1/tfl/lines` response - `route_variants` should contain hub codes like `"HUBPAD"` instead of raw NaPTAN IDs like `"940GZZLUPAC"`
4. Test frontend - select a hub station (e.g., Paddington) and verify available trains/destinations are shown

## Files Changed

| File | Changes | Lines |
|------|---------|-------|
| `backend/alembic/versions/55ac4b0569a1_*.py` | Add migration | +37 |
| `backend/app/helpers/station_resolution.py` | Add translation helpers | +72 |
| `backend/app/models/tfl.py` | Add `route_variants_canonical` field | +6 |
| `backend/app/schemas/tfl.py` | Update LineResponse | +4/-2 |
| `backend/app/services/tfl_service.py` | Fix cache key, compute canonical | +37/-3 |
| `backend/tests/test_station_resolution_helpers.py` | Add tests | +169 |
| **Total** | | **+327/-3** |

## Breaking Changes

None. All changes are backward compatible:
- Backend validation continues using raw `route_variants` 
- Frontend receives canonical version but via same API field name
- No API contract changes

## Summary by Sourcery

Ensure hub stations correctly expose route variants with canonical hub IDs and invalidate the appropriate lines cache when rebuilding the station graph.

New Features:
- Store canonical route variants per line using hub-based station identifiers for API responses.

Bug Fixes:
- Fix cache invalidation to target the correct lines cache key used by line-fetching logic, preventing stale route data after graph rebuilds.
- Align route variant station IDs with canonical hub IDs so hub stations correctly match frontend station identifiers and display trains.

Enhancements:
- Introduce helpers to map NaPTAN IDs to canonical station IDs and translate route variants accordingly, computed as part of the graph build workflow.

Build:
- Add a database migration to add the route_variants_canonical JSON column to the lines table.

Tests:
- Add unit tests covering station-to-canonical mapping and translation of route variants to canonical IDs, including edge cases.

### CODE REVIEW

File: backend/app/services/tfl_service.py [hunk 1]
```diff
@@ -2792,6 +2794,36 @@ def _store_line_routes(
...
+    async def _compute_canonical_route_variants(self, lines: list[Line]) -> None:
+        """
+        Compute canonical route variants with hub codes for API responses.
+
+        Translates NaPTAN IDs in route_variants to canonical station IDs
+        (hub codes where applicable) and stores in route_variants_canonical.
+
+        This is done after all stations have been fetched and stored,
+        so the naptan-to-canonical mapping is complete.
+
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py:2806]: Fixed in bddbba2. Added else clause to clear route_variants_canonical when route_variants is None/empty.

File: backend/app/schemas/tfl.py [hunk 2]
```diff
@@ -39,7 +39,11 @@ class LineResponse(BaseModel):
     tfl_id: str
     name: str
     mode: str  # Transport mode: "tube", "overground", "dlr", "elizabeth-line", etc.
-    route_variants: RoutesData | None = None  # Route sequences for branch-aware validation
+    route_variants: RoutesData | None = Field(
+        default=None,
+        validation_alias="route_variants_canonical",  # Read from canonical field (with hub IDs)
+        description="Route sequences for branch-aware validation (with canonical station IDs)",
+    )
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/schemas/tfl.py]: Fixed in bddbba2. Changed validation_alias to serialization_alias. The field is now named route_variants_canonical in the schema and uses serialization_alias='route_variants' to maintain API contract.

File: backend/app/services/tfl_service.py [hunk 3]
```diff
@@ -2792,6 +2794,36 @@ def _store_line_routes(
...
+            if line.route_variants:
+                line.route_variants_canonical = translate_route_variants_to_canonical(
+                    line.route_variants, naptan_to_canonical
+                )
+                logger.debug(
+                    "translated_route_variants_to_canonical",
+                    line_tfl_id=line.tfl_id,
+                    route_count=len(line.route_variants.get("routes", [])),
+                )
+
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py:2828]: Fixed in bddbba2. Added comprehensive test coverage:
- test_compute_canonical_route_variants_translates_naptan_to_hub
- test_compute_canonical_route_variants_clears_when_route_variants_none
- test_compute_canonical_route_variants_handles_empty_stations
- test_get_lines_returns_canonical_route_variants (integration test)

All tests use the test network per project guidelines.

**Comment by mnbf9rca** [comment]
## Review Comments Addressed

All review comments have been resolved in commit bddbba2:

### Issues Fixed
1. ✅ **Critical Pydantic Bug** - Changed `validation_alias` to `serialization_alias` in LineResponse
2. ✅ **Stale Data Bug** - Added else clause to clear `route_variants_canonical` when `route_variants` is None
3. ✅ **Test Coverage** - Added 5 comprehensive tests covering all edge cases

### Test Results
- All 1620 tests passed
- Coverage: 95.97% (exceeds 95% requirement)
- All pre-commit hooks passed

### Tests Added
- `test_multiple_naptan_children_same_hub` - Multiple NaPTAN children → same hub
- `test_compute_canonical_route_variants_translates_naptan_to_hub` - Basic translation
- `test_compute_canonical_route_variants_clears_when_route_variants_none` - Stale data clearing
- `test_compute_canonical_route_variants_handles_empty_stations` - Edge case handling
- `test_get_lines_returns_canonical_route_variants` - Integration test

All tests updated to use test network stations per project guidelines.
